packages/vite/src/node/plugins/define.ts

```diff
@@ -145,36 +135,88 @@ export function definePlugin(config: ResolvedConfig): Plugin {
         return
       }
 
-      const [replacements, pattern] = ssr ? ssrPattern : defaultPattern
+      const [define, pattern] = ssr ? ssrPattern : defaultPattern
+      if (!pattern) return
 
-      if (!pattern) {
-        return null
-      }
+      // Check if our code needs any replacements before running esbuild
+      pattern.lastIndex = 0
+      if (!pattern.test(code)) return
 
-      if (ssr && !isBuild) {
-        // ssr + dev, simple replace
-        return code.replace(pattern, (_, match) => {
-          return '' + replacements[match]
-        })
-      }
+      return await replaceDefine(code, id, define, config)
+    },
+  }
+}
 
-      const s = new MagicString(code)
-      let hasReplaced = false
-      let match: RegExpExecArray | null
+export async function replaceDefine(
+  code: string,
+  id: string,
+  define: Record<string, string>,
+  config: ResolvedConfig,
+): Promise<{ code: string; map: string | null }> {
+  // Because esbuild only allows JSON-serializable values, and `import.meta.env`
+  // may contain values with raw identifiers, making it non-JSON-serializable,
+  // we replace it with a temporary marker and then replace it back after to
+  // workaround it. This means that esbuild is unable to optimize the `import.meta.env`
+  // access, but that's a tradeoff for now.
+  const replacementMarkers: Record<string, string> = {}
+  const env = define['import.meta.env']
+  if (env && !canJsonParse(env)) {
+    const marker = `_${getHash(env, env.length - 2)}_`
+    replacementMarkers[marker] = env
+    define = { ...define, 'import.meta.env': marker }
+  }
 
-      while ((match = pattern.exec(code))) {
-        hasReplaced = true
-        const start = match.index
-        const end = start + match[0].length
-        const replacement = '' + replacements[match[1]]
-        s.update(start, end, replacement)
-      }
+  const esbuildOptions = config.esbuild || {}
 
-      if (!hasReplaced) {
-        return null
-      }
+  const result = await transform(code, {
+    loader: 'js',
+    charset: esbuildOptions.charset ?? 'utf8',
+    platform: 'neutral',
+    define,
+    sourcefile: id,
+    sourcemap: config.command === 'build' && !!config.build.sourcemap,
```

Comment:
I think we always need to generate the sourcemap because esbuild will trim empty lines and comments.
https://esbuild.github.io/try/#dAAwLjE5LjUAewogICAgbG9hZGVyOiAnanMnLAogICAgY2hhcnNldDogJ3V0ZjgnLAogICAgcGxhdGZvcm06ICduZXV0cmFsJywKICAgIGRlZmluZTogeyAiaW1wb3J0Lm1ldGEuZW52LlZBTFVFIjogIlwiZm9vYmFyXCIiIH0KfQBjb25zdCBmb28gPQogIGltcG9ydC5tZXRhLmVudi5WQUxVRQoKLyogYmF6ICovCgovLyBmb29iYXIKY29uc29sZS5sb2coZm9vKQ